### PR TITLE
Fixes for One-Shot MPT-7b

### PR DESCRIPTION
**Pull request by Satrat**
The one-shot script had previously been untested on MPT, this PR fixes the script to work with MPT

### Summary of Changes
* Updated `cache_attention_inputs` to support the `model.transformers.blocks` format of MPT. This requires a change to how `layer_prefix` is defined in recipe metadata
* `cache_attention_inputs` now searches for the embedding layer instead of referencing it by a hardcoded name
* Added `position_bias` to `DEFAULT_TARGET_IDS`
* `model.seqlen` default updated to work with MPT

### Testing

test_mpt_recipe.yaml
```
metadata:
  target_model:
    layer_prefix: "transformer.blocks"
    architecture: "mpt"
test_stage:
  obcq_modifiers:
    SparseGPTModifier:
      sparsity: 0.5
      block_size: 128
      sequential_update: False
      quantize: False
      percdamp: 0.01
      prunen: 0
      prunem: 0
      targets: [
        "re:transformer.blocks.\\d+$"
      ]
```

Launch OBCQ on MPT-7b
```
python src/sparseml/transformers/sparsification/obcq/obcq.py mosaicml/mpt-7b open_platypus --recipe test_mpt_recipe.yaml --precision float16 --save True
```

Asana ticket: https://app.asana.com/0/1205229323407165/1206141089429597/f

### CODE REVIEW

File: src/sparseml/modifiers/obcq/utils/helpers.py [hunk 1]
```diff
@@ -127,24 +126,36 @@ def execute_offloaded_module(
...
     model, dataloader, device, nsamples, target_ids, layer_prefix
 ):
-    if layer_prefix:
-        embed_tokens = getattr(model.model, layer_prefix).embed_tokens
-        first_layer = getattr(model.model, layer_prefix).layers[0]
+    if layer_prefix:  # get model-specific path to layers list
+        split_prefix = layer_prefix.split(".")
+        layers_name = split_prefix[-1]
+        model_root_name = ".".join(split_prefix[:-1])
+        model_root = operator.attrgetter(model_root_name)(model)
```
Comment (rahul-tuli) [inline src/sparseml/modifiers/obcq/utils/helpers.py:133]: nice!

**Review by rahul-tuli — APPROVED** [review]
LGTM!

**Review by bfineran — APPROVED** [review]

**Review by mwitiderrick — CHANGES_REQUESTED** [review]
@Satrat I ran the given command with the given recipe on this branch and got the following error 
```
2023-12-13 01:35:08 __main__     WARNING  A supported model type(['opt', 'llama', 'mistral']) could not be parsed from model_path=mosaicml/mpt-7b. Defaulting to AutoModelForCausalLM loading.
Loading checkpoint shards: 100%|████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.72s/it]
Traceback (most recent call last):
  File "pydantic/main.py", line 522, in pydantic.main.BaseModel.parse_obj
ValueError: dictionary update sequence element #0 has length 1; 2 is required

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/michael/mwitiderrick/sparseml/src/sparseml/transformers/sparsification/obcq/obcq.py", line 240, in <module>
    one_shot(
  File "/home/michael/mwitiderrick/sparseml/src/sparseml/transformers/sparsification/obcq/obcq.py", line 137, in one_shot
    session.apply(
  File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/session.py", line 235, in apply
    self.initialize(**kwargs)
  File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/session.py", line 181, in initialize
    mod_data = self._lifecycle.initialize(
  File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/lifecycle/session.py", line 88, in initialize
    extras = self.recipe_container.update(**extras)
  File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/recipe/container.py", line 116, in update
    rec = Recipe.create_instance(rec)
  File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/recipe/recipe.py", line 76, in create_instance
    return Recipe.parse_obj(obj)
  File "pydantic/main.py", line 525, in pydantic.main.BaseModel.parse_obj
pydantic.error_wrappers.ValidationError: 1 validation error for Recipe
__root__
  Recipe expected dict not str (type=type_error)

```

Tried twice, same error

**Comment by Satrat** [comment]
> @Satrat I ran the given command with the given recipe on this branch and got the following error
> 
> ```
> 2023-12-13 01:35:08 __main__     WARNING  A supported model type(['opt', 'llama', 'mistral']) could not be parsed from model_path=mosaicml/mpt-7b. Defaulting to AutoModelForCausalLM loading.
> Loading checkpoint shards: 100%|████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.72s/it]
> Traceback (most recent call last):
>   File "pydantic/main.py", line 522, in pydantic.main.BaseModel.parse_obj
> ValueError: dictionary update sequence element #0 has length 1; 2 is required
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/home/michael/mwitiderrick/sparseml/src/sparseml/transformers/sparsification/obcq/obcq.py", line 240, in <module>
>     one_shot(
>   File "/home/michael/mwitiderrick/sparseml/src/sparseml/transformers/sparsification/obcq/obcq.py", line 137, in one_shot
>     session.apply(
>   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/session.py", line 235, in apply
>     self.initialize(**kwargs)
>   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/session.py", line 181, in initialize
>     mod_data = self._lifecycle.initialize(
>   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/lifecycle/session.py", line 88, in initialize
>     extras = self.recipe_container.update(**extras)
>   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/recipe/container.py", line 116, in update
>     rec = Recipe.create_instance(rec)
>   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/recipe/recipe.py", line 76, in create_instance
>     return Recipe.parse_obj(obj)
>   File "pydantic/main.py", line 525, in pydantic.main.BaseModel.parse_obj
> pydantic.error_wrappers.ValidationError: 1 validation error for Recipe
> __root__
>   Recipe expected dict not str (type=type_error)
> ```
> 
> Tried twice, same error

Hey @mwitiderrick, this looks like there was an error loading the recipe file. Could you double check the filename you're passing as a recipe is correct? I just double checked with the recipe above and the script is working:
```
src/sparseml/transformers/sparsification/obcq/obcq.py mosaicml/mpt-7b open_platypus --recipe test_mpt_recipe.yaml 
```

**Review by mwitiderrick — APPROVED** [review]

**Comment by mwitiderrick** [comment]
> > @Satrat I ran the given command with the given recipe on this branch and got the following error
> > ```
> > 2023-12-13 01:35:08 __main__     WARNING  A supported model type(['opt', 'llama', 'mistral']) could not be parsed from model_path=mosaicml/mpt-7b. Defaulting to AutoModelForCausalLM loading.
> > Loading checkpoint shards: 100%|████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.72s/it]
> > Traceback (most recent call last):
> >   File "pydantic/main.py", line 522, in pydantic.main.BaseModel.parse_obj
> > ValueError: dictionary update sequence element #0 has length 1; 2 is required
> > 
> > The above exception was the direct cause of the following exception:
> > 
> > Traceback (most recent call last):
> >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/transformers/sparsification/obcq/obcq.py", line 240, in <module>
> >     one_shot(
> >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/transformers/sparsification/obcq/obcq.py", line 137, in one_shot
> >     session.apply(
> >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/session.py", line 235, in apply
> >     self.initialize(**kwargs)
> >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/session.py", line 181, in initialize
> >     mod_data = self._lifecycle.initialize(
> >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/lifecycle/session.py", line 88, in initialize
> >     extras = self.recipe_container.update(**extras)
> >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/recipe/container.py", line 116, in update
> >     rec = Recipe.create_instance(rec)
> >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/recipe/recipe.py", line 76, in create_instance
> >     return Recipe.parse_obj(obj)
> >   File "pydantic/main.py", line 525, in pydantic.main.BaseModel.parse_obj
> > pydantic.error_wrappers.ValidationError: 1 validation error for Recipe
> > __root__
> >   Recipe expected dict not str (type=type_error)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Tried twice, same error
> 
> Hey @mwitiderrick, this looks like there was an error loading the recipe file. Could you double check the filename you're passing as a recipe is correct? I just double checked with the recipe above and the script is working:
> 
> ```
> src/sparseml/transformers/sparsification/obcq/obcq.py mosaicml/mpt-7b open_platypus --recipe test_mpt_recipe.yaml 
> ```

@Satrat this sounds like an opportunity to improve this error message. 

We should tell the developer that the recipe can't be found by checking if that file exists in the given path.

**Comment by Satrat** [comment]
> > > @Satrat I ran the given command with the given recipe on this branch and got the following error
> > > ```
> > > 2023-12-13 01:35:08 __main__     WARNING  A supported model type(['opt', 'llama', 'mistral']) could not be parsed from model_path=mosaicml/mpt-7b. Defaulting to AutoModelForCausalLM loading.
> > > Loading checkpoint shards: 100%|████████████████████████████████████████████████| 2/2 [00:07<00:00,  3.72s/it]
> > > Traceback (most recent call last):
> > >   File "pydantic/main.py", line 522, in pydantic.main.BaseModel.parse_obj
> > > ValueError: dictionary update sequence element #0 has length 1; 2 is required
> > > 
> > > The above exception was the direct cause of the following exception:
> > > 
> > > Traceback (most recent call last):
> > >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/transformers/sparsification/obcq/obcq.py", line 240, in <module>
> > >     one_shot(
> > >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/transformers/sparsification/obcq/obcq.py", line 137, in one_shot
> > >     session.apply(
> > >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/session.py", line 235, in apply
> > >     self.initialize(**kwargs)
> > >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/session.py", line 181, in initialize
> > >     mod_data = self._lifecycle.initialize(
> > >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/lifecycle/session.py", line 88, in initialize
> > >     extras = self.recipe_container.update(**extras)
> > >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/recipe/container.py", line 116, in update
> > >     rec = Recipe.create_instance(rec)
> > >   File "/home/michael/mwitiderrick/sparseml/src/sparseml/core/recipe/recipe.py", line 76, in create_instance
> > >     return Recipe.parse_obj(obj)
> > >   File "pydantic/main.py", line 525, in pydantic.main.BaseModel.parse_obj
> > > pydantic.error_wrappers.ValidationError: 1 validation error for Recipe
> > > __root__
> > >   Recipe expected dict not str (type=type_error)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Tried twice, same error
> > 
> > 
> > Hey @mwitiderrick, this looks like there was an error loading the recipe file. Could you double check the filename you're passing as a recipe is correct? I just double checked with the recipe above and the script is working:
> > ```
> > src/sparseml/transformers/sparsification/obcq/obcq.py mosaicml/mpt-7b open_platypus --recipe test_mpt_recipe.yaml 
> > ```
> 
> @Satrat this sounds like an opportunity to improve this error message.
> 
> We should tell the developer that the recipe can't be found by checking if that file exists in the given path.
> 
> We can also add a log message telling the user the path to the recipe being used for one-shot.

Yes, the reason this error message pops up is because if we cant parse the input string as a filename, we try to parse it as a raw string of yaml, because we support passing the recipe that way as well. Agreed that a warning log would be helpful here to indicate the file couldn't be found, I'll put that in a separate branch though since its unrelated to MPT

**Pull request merged**